### ISSUE TITLE
Deploying cluster autoscaler but not enabling any node pools results in confusing logs and access to non-kube ASGs

### ISSUE THREAD

**Issue by c-knowles** [kubernetes-incubator/kube-aws#800]
As of 61e411229d5b6fa6c0fbbec1c0e3e91952a4c053, turning on the cluster autoscaler with RBAC enabled gets logs similar to the following:

```W  Failed to find readiness information for k8s-preprod-v7-Controlplane-1VFZCIM8O6ZUO-Controllers
W  Failed to find readiness information for k8s-dev-v8-PoolC-1JBFUB0G59YY7-Workers 
W  Failed to find readiness information for k8s-dev-v8-Controlplane-1ERY9ADUQZRDI-Etcd0
```

I've only deployed the autoscaler to a single cluster (dev) but there are also the same logs for another totally different cluster in the same AWS account.

**Comment by mumoshu** [comment]
@c-knowles Thanks for the feedback!

I can't recall whether the messages had been there or not when I first developed the feature on CA.

For the purpose of gathering more information on investigation, would you mind confirming:
* Did you enable autoscaling by `worker.nodePools[].autoscaling.clusterAutoscaler.enabled`?
* Does autoscaling(both scaling-up and down) work or not?

**Comment by c-knowles** [comment]
@mumoshu I can confirm that the autoscaler app is deployed. I have not enabled it on any pools yet however I did this as a first step to see it deployed and thought it should be doing nothing rather than logging warnings about unrelated ASGs. I'm also getting logs about ASGs unrelated to k8s/kube-aws which is a bit concerning. Maybe the autoscaler needs to filter prior to logging about unrelated ASGs.

**Comment by c-knowles** [comment]
Just confirmed if I enable on a single node pool that all the warnings disappear. I suppose this is mostly unexpected behaviour then when it's not activated but why is the autoscaler even concerned about those other ASGs at all?

**Comment by mumoshu** [comment]
@c-knowles Thanks for the confirmation 👍 

AFAIK, CA should [try to get readiness information only when there are one or more node groups](https://github.com/kubernetes/autoscaler/blob/591ced9c7c61ec3ccdedded46f75e1be1856016d/cluster-autoscaler/clusterstate/clusterstate.go#L237).
Node groups in CA are backed by Spot Fleets or ASGs in kube-aws. Spot Fleets are not supported by CA though).

Perhaps there is a bug in CA that CA tries to get readiness information on an unknown node group when there are no node groups configured as scaling targets?

**Comment by c-knowles** [comment]
I had a look through the auto scaler code and can't spot anything obvious. I guess I should report these, as you can see it starts to query some nodes and I had not even switched it on for any ASGs. i.e. no ASGs tagged appropriately.

```
I  Registered cleanup signal handler 
I  successfully renewed lease kube-system/cluster-autoscaler 
I  successfully renewed lease kube-system/cluster-autoscaler 
I  successfully renewed lease kube-system/cluster-autoscaler 
I  successfully renewed lease kube-system/cluster-autoscaler 
I  successfully renewed lease kube-system/cluster-autoscaler 
I  successfully renewed lease kube-system/cluster-autoscaler 
I  Poll finished 
I  successfully renewed lease kube-system/cluster-autoscaler 
I  Regenerating ASG information for non-kube-asg-1 
I  Regenerating ASG information for non-kube-asg-2 
I  Regenerating ASG information for k8s-dev-v8-Controlplane-1ERY9ADUQZRDI-Controllers 
I  Regenerating ASG information for k8s-dev-v8-Controlplane-1ERY9ADUQZRDI-Etcd0 
I  Regenerating ASG information for k8s-dev-v8-PoolA-YUMGXD2TB6JS-Workers 
I  Regenerating ASG information for k8s-dev-v8-PoolB-16QTO48LELF2E-Workers 
I  Regenerating ASG information for k8s-dev-v8-PoolC-1JBFUB0G59YY7-Workers 
I  Regenerating ASG information for k8s-preprod-v7-Controlplane-1VFZCIM8O6ZUO-Controllers 
I  Regenerating ASG information for k8s-preprod-v7-Controlplane-1VFZCIM8O6ZUO-Etcd0 
I  Regenerating ASG information for k8s-preprod-v7-Controlplane-1VFZCIM8O6ZUO-Etcd1 
I  Regenerating ASG information for k8s-preprod-v7-Controlplane-1VFZCIM8O6ZUO-Etcd2 
I  Regenerating ASG information for k8s-preprod-v7-PoolA-I72A3JMHE9BG-Workers 
I  Regenerating ASG information for k8s-preprod-v7-PoolB-16IM7SE348HFH-Workers 
I  Regenerating ASG information for k8s-preprod-v7-PoolC-BOV3E4H9622M-Workers 
W  Readiness for node group non-kube-asg-1 not found 
W  Readiness for node group non-kube-asg-2 not found 
W  Readiness for node group k8s-dev-v8-Controlplane-1ERY9ADUQZRDI-Etcd0 not found 
W  Readiness for node group k8s-dev-v8-PoolC-1JBFUB0G59YY7-Workers not found 
W  Readiness for node group k8s-preprod-v7-Controlplane-1VFZCIM8O6ZUO-Controllers not found 
W  Readiness for node group k8s-preprod-v7-Controlplane-1VFZCIM8O6ZUO-Etcd0 not found 
W  Readiness for node group k8s-preprod-v7-Controlplane-1VFZCIM8O6ZUO-Etcd1 not found 
W  Readiness for node group k8s-preprod-v7-Controlplane-1VFZCIM8O6ZUO-Etcd2 not found 
W  Readiness for node group k8s-preprod-v7-PoolA-I72A3JMHE9BG-Workers not found 
W  Readiness for node group k8s-preprod-v7-PoolB-16IM7SE348HFH-Workers not found 
W  Readiness for node group k8s-preprod-v7-PoolC-BOV3E4H9622M-Workers not found 
I  21 unregistered nodes present 
I  Filtering out schedulables 
I  No schedulable pods 
I  No unschedulable pods 
I  Scale down status: unneededOnly=true lastScaleUpTime=2017-07-21 02:03:03.834542876 +0000 UTC lastScaleDownFailedTrail=2017-07-21 02:03:03.834543236 +0000 UTC schedulablePodsPresent=false 
I  Calculating unneeded nodes 
I  Node ip-10-0-30-221.eu-west-1.compute.internal - utilization 0.335000 
I  Node ip-10-0-31-193.eu-west-1.compute.internal - utilization 0.561000 
```

**Comment by joseppla** [comment]
Seems the problem's bigger than seems, autoscaler started to change the size of ASGs that are not part of the k8s cluster!!!

**Comment by c-knowles** [comment]
While it would not solve any possible root cause there, I think we should also look at updating the IAM resources allowed so it's a list of the relevant node pool ASGs rather than `*`:
https://github.com/kubernetes-incubator/kube-aws/blob/a2daf1a95a40a6326f38fb1c244b0d8a6b790289/core/controlplane/config/templates/stack-template.json#L246

**Comment by mumoshu** [comment]
@joseppla Hi! I'm interested. Could you share me a full output of `kubectl logs` for the cluster-autoscaler pod, your cluster.yaml, and kube-aws version?

**Comment by joseppla** [comment]
@mumoshu Hi , I deleted the cluster but fortunately still have the terminal output :) the version was the master branch, not sure if from yesterday or from this morning, I still have the cloned repo so I'm gonna build it again and tell you the exact version, but later, now I must go, I have all the CA logs and the cluster.yaml, I'll post it all in 3 or 4 hours.

**Comment by joseppla** [comment]
@mumoshu Hi again, attached part of the CA log output, not all because is very large and it's always the same repeated messages, I removed some info "company data protection", but think with this you'll have enough logs to see whats happening and the important messages. The building message shows 'Building kube-aws a5170e00705c094c0d5bee956a11dd3894def479'. Attached you'll find the cluster.yaml  too.
[ca-logs.txt](https://github.com/kubernetes-incubator/kube-aws/files/1261859/ca-logs.txt)
[cluster.txt](https://github.com/kubernetes-incubator/kube-aws/files/1261877/cluster.txt)

**Comment by mumoshu** [comment]
@joseppla Thanks for the info 👍 

Unfortunately, I'm unable to reproduce the problem.
The messages like `Regenerating ASG information for <name of asg seemingly unrelated to k8s>`  does indicate that you had by any chance provided an incorrect parameter to CA's `--node-group-auto-discovery` option. However, I don't think you did that.

To investigate further:

* Would you mind providing me a result of `kubectl get deployment cluster-autoscaler --namespace kube-system`?
* Also, would you mind running CA with `--v=6` instead of the default `--v=4`, kill the CA pod by setting replicas to 0 before it acts badly, and finally share me a result of `kubectl logs` for the CA pod?

**Comment by mumoshu** [comment]
@c-knowles Would you mind providing me the same information as above for further investigation?

**Comment by joseppla** [comment]
@mumoshu Hi again, attached part of the CA log output, not all because is very large and it's always the same repeated messages, I removed some info "company data protection", but think with this you'll have enough logs to see whats happening and the important messages. The building message shows 'Building kube-aws a5170e00705c094c0d5bee956a11dd3894def479'. Attached you'll find the cluster.yaml  too.
[ca-logs.txt](https://github.com/kubernetes-incubator/kube-aws/files/1261859/ca-logs.txt)
[cluster.txt](https://github.com/kubernetes-incubator/kube-aws/files/1261877/cluster.txt)

**Comment by joseppla** [comment]
@mumoshu Hi, I didn't make any change on the CA deployment config, was the default automated deployment made by kube-aws, unfortunately I don't have the cluster running with this version, after the small disaster I deleted the cluster and rolled back to a stable version of the kube-aws so I can't recover the deployment, a pity because yesterday I had it, I modified the deployment fixing the autodiscover parameters with fixed asg names :( . The only thing I can do is deploy the same config again on an isolated vpc and get the deployment and the logs but I need some time to create fake ASG groups to check if we can reproduce the problem.

**Comment by mumoshu** [comment]
@c-knowles @joseppla Thanks for your patience. I guess your issue is https://github.com/kubernetes/autoscaler/issues/278#issuecomment-325959675.
WDYT?

@joseppla Probably you missed registering one or more node pools as autoscaling targets?

The trick, or maybe 
 trap, is that you have to set `worker.nodePools[].autoscaling.clusterAutoscaler.enabled` to actually register the node pool to CA. Otherwise, no node pools/ASGs are detected by CA. Accoding to your cluster.yaml, you do have `worker.nodePools[].clusterAutoscaler.enabled`, which misses `autoscaling`(kube-aws should emit a validation error for the unexpected config key but it is an another story :smile:

If that's the case, your issue will go away once you register one or more targets, anyway.

The CA bug which results in terminating nodes outside of k8s clusters does exist, but happens only when you have no scaling targets. It will be addressed in CA as discussed in https://github.com/kubernetes/autoscaler/issues/278#issuecomment-325959675.

**Comment by mumoshu** [comment]
@c-knowles 

> While it would not solve any possible root cause there, I think we should also look at updating the IAM resources allowed so it's a list of the relevant node pool ASGs rather than *

Although I'm unsure how we could limit the ASG listing, we already have a set of IAM permissions to restrict worker role to only allow terminating instances relevant to the cluster.
The problem is that we missed the restriction in the controller role, which is going to be fixed via #903

**Comment by c-knowles** [comment]
@mumoshu yes it looks like my issue was https://github.com/kubernetes/autoscaler/issues/278#issuecomment-325959675. For the restrictions, that's cool at least it can't update them now.

**Comment by c-knowles** [comment]
@mumoshu BTW, I think you are right - `autoscaling:DescribeAutoScalingGroups` only supports `*` resource and no conditions.

**Comment by joseppla** [comment]
@mumoshu sorry for the delay, true, I've missed part of the parameter.

**Comment by adimarco** [comment]
This is... really dangerous.  Really, really dangerous.

In my case, I was going through `cluster.yaml` and turning on autoscaling, and got to the bit in my nodepool configuration that looks like this:

```
#      # Autoscaling by adding/removing nodes according to resource usage
#      autoscaling:
#        # Make this node pool an autoscaling-target of k8s cluster-autoscaler
#        #
#        # Beware that making this an autoscaing-target doesn't automatically deploy cluster-autoscaler itself -
#        # turn on `addons.clusterAutoscaler.enabled` to deploy it on controller nodes.
#        clusterAutoscaler:
#          enabled: true
```
I uncommented these two lines for my nodepool without noticing the lone `autoscaling:` tag nestled in the block of comments above it:
```
#        clusterAutoscaler:
#          enabled: true
```
Got a complaint about tags in the wrong place, figured I'd fat-fingered the space bar and messed up the yaml indentation in the middle of that mess of comment tags, pulled those lines up by 2 spaces, and everything deployed fine.

A few minutes later my all of my non-kubernetes production app server ASGs (in a completely different VPC) just suddenly started scaling down because cluster-autoscaler considers their instances unregistered.  It only stopped when it hit the ASG min and couldn't terminate any more.

```
W0921 19:31:33.346253       1 clusterstate.go:397] Readiness for node group REDACTED-1 not found
W0921 19:31:33.346256       1 clusterstate.go:397] Readiness for node group REDACTED-2 not found
I0921 19:31:33.346331       1 static_autoscaler.go:145] 64 unregistered nodes present
I0921 19:31:33.346338       1 utils.go:285] Removing unregistered node aws:///us-east-1b/i-REDACTED
I0921 19:31:33.557700       1 aws_manager.go:147] Terminating EC2 instance: i-REDACTED
I0921 19:31:33.557721       1 utils.go:285] Removing unregistered node aws:///us-east-1d/i-REDACTED
I0921 19:31:33.824472       1 aws_manager.go:147] Terminating EC2 instance: i-REDACTED
```

etc.

I get that I deployed this change myself, but the default behavior for a small configuration mistake like that should really be something other than "just start terminating everything in all ASGs"

**Comment by mumoshu** [comment]
@adimarco Sorry for the trouble. You are absolutely right.

The problem itself is fixed via #903 and configuration mistakes will be prevented via #905.
For #903, it is now impossible for CA to delete ASGs outside of the cluster. This is enforced by AWS IAM.

Both fixes are going to be included in v0.9.9-rc.1.

**Comment by mumoshu** [comment]
Closing this as fixed for now. However, please feel free to reopen if necessary.
